iOS SDK Swift migration (#506)
rudderlabs/rudderstack-docs: stream-sources/rudderstack-sdk-integration-guides/rudderstack-ios-sdk/ios-v2.mdx

---
# slug: "/docs/stream-sources/rudderstack-sdk-integration-guides/rudderstack-ios-sdk"

> title: "iOS"

Reviewer: Should be `Swift`.

description: >-

>   Detailed technical documentation on the RudderStack iOS SDK to send

Reviewer: ```suggestion
  Detailed technical documentation on the RudderStack SDK to send
```

>   events from your iOS apps to various destinations.

Reviewer: ```suggestion
  events from your iOS/tvOS/watchOS/macOS apps to various destinations.
```

---

# iOS

The **RudderStack iOS SDK** lets you track your customer event data from your iOS applications and send it to your specified destinations via RudderStack.

Check out the [GitHub codebase](https://github.com/rudderlabs/rudder-sdk-ios) to get a more hands-on understanding of the SDK.

<div class="successBlock">

You can now integrate the iOS SDK with your tvOS and watchOS apps and seamlessly track user events without the need for any additional configuration. The iOS SDK supports tvOS starting from version <strong>1.1.0</strong> and <strong>watchOS</strong> is supported starting from the SDK version <strong>1.3.1</strong>.
</div>

// Add stability beta badge here

## SDK setup requirements

> To set up the RudderStack iOS SDK, the following prerequisites must be met:

Reviewer: ```suggestion
To set up the RudderStack Swift SDK, the following prerequisites must be met:
```

- You will need to set up a [RudderStack account](https://app.rudderstack.com).
- Set up an iOS source in the dashboard. For more information, follow the [Adding a Source and Destination in RudderStack](https://rudderstack.com/docs/connections/adding-source-and-destination-rudderstack/) guide.  You should then see a **Write Key** for this source, as shown:

<img src="../../../assets/event-stream-sources/ios-sdk-1.png" alt="iOS source write key" />

- You also need a data plane URL. Follow [**this section**](https://rudderstack.com/docs/rudderstack-open-source/installing-and-setting-up-rudderstack/#what-is-a-data-plane-url-where-do-i-get-it) for more information on the data plane URL and where to get it.
- Finally, you will need a Mac with the latest version of [Xcode](https://developer.apple.com/xcode/).

> ## Installing the RudderStack iOS SDK

Reviewer: ```suggestion
## Installing the RudderStack Swift SDK
```

> The RudderStack iOS SDK is distributed through [Cocoapods](https://cocoapods.org/pods/Rudder) and [Carthage](https://github.com/Carthage/Carthage).

Reviewer: ```suggestion
The RudderStack SDK is distributed through [Cocoapods](https://cocoapods.org/pods/Rudder) and [Carthage](https://github.com/Carthage/Carthage).
```

<div class="successBlock">

The recommended and easiest way to add the iOS SDK to your project is through <code class="inline-code">Podfile</code>.
</div>

> To install the iOS SDK depending on your preference, follow the steps below:

Reviewer: ```suggestion
To install the SDK depending on your preference, follow the steps below:
```

<Tabs>
  <TabList>
    <Tab>CocoaPods</Tab>

>     <Tab>Carthage</Tab>

Author: Will the Carthage installation steps remain the same?

  </TabList>
  <TabPanels>
    <TabPanel>
      <ul>

>       <li>Add the iOS SDK to your <code class="inline-code">Podfile</code>, as shown:

Reviewer: ```suggestion
      <li>Add the SDK to your <code class="inline-code">Podfile</code>, as shown:
```

<span>

```ruby

> pod 'RudderStack', '~> 1.0.0-beta.3'

Reviewer: ```suggestion
pod 'Rudder', '~> 2.0.0'
```

```
</span>
      </li>
      <li>Then, run the following command:
<span>

```bash
pod install
```
</span>
      </li>
      </ul>
    </TabPanel>
    <TabPanel>
      <ul>
      <li>Add the RudderStack SDK to your <code class="inline-code">Cartfile</code>, as shown:
<span>

```ruby
github "rudderlabs/rudder-sdk-ios"
```
</span>
      </li>
      <li>Then, run the following command:
<span>

```bash
carthage update
```
</span>
      </li>
      </ul>
    </TabPanel>
  </TabPanels>
</Tabs>

Remember to include the following code in all the `.m` and `.h` files (Objective-C) or the `.swift` files where you want to refer to or use the RudderStack SDK classes:

<Tabs>
  <TabList>
    <Tab>Objective-C</Tab>
    <Tab>Swift</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```objectivec

> #import <Rudder/Rudder.h>

Reviewer: ```suggestion
@import Rudder;
```

```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
import Rudder
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

<div class="dangerBlock">

RudderStack uses <a href="https://sqlite.org/index.html">SQLite</a> to temporarily store the events before sending them to the backend (data plane). Making calls which are not thread-safe, like <code class="inline-code">SQLite.shutdown()</code>, might lead to unexpected crashes.
</div>

## Initializing the RudderStack client

To initialize the RudderStack client, place the following code in your `AppDelegate.m` file under the `didFinishLaunchingWithOptions` method:

<Tabs>
  <TabList>
    <Tab>Objective-C</Tab>
    <Tab>Swift</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```objectivec
RSConfig *config = [[RSConfig alloc] initWithWriteKey:WRITE_KEY];
[config dataPlaneURL:DATA_PLANE_URL];
[config trackLifecycleEvents:YES];
[config recordScreenViews:YES];

> [[RSClient sharedInstance] configureWith:config]

Reviewer: ```suggestion
[[RSClient sharedInstance] configureWith:config];
```

                  .trackLifecycleEvents(true)
                  .recordScreenViews(true)
        
RSClient.sharedInstance().configure(with: config)
```
</span>

A shared instance of <code class="inline-code">RSClient</code> is accesible after the initialization by <code class="inline-code">RSClient.sharedInstance()</code>.
      </TabPanel>
    </TabPanels>
</Tabs>

<div class="infoBlock">

RudderStack automatically tracks the following <a href="https://www.rudderstack.com/docs/rudderstack-api/api-specification/application-lifecycle-events-spec/">application lifecycle events</a>:
<ul>
<li><code class="inline-code">Application Installed</code></li>
<li><code class="inline-code">Application Updated</code></li>
<li><code class="inline-code">Application Opened</code></li>
<li><code class="inline-code">Application Backgrounded</code></li>
</ul>
You can disable these events using the <code class="inline-code">trackLifecycleEvents</code> method of <code class="inline-code">RSConfig</code> and passing <code class="inline-code">false</code>. However, it is highly recommended to keep them enabled.
</div>

## Configuring the RudderStack client

> You can configure your client based on the following parameters using `RudderConfig`:

Reviewer: ```suggestion
You can configure your client based on the following parameters using `RSConfig`:
```

| Parameter               | Type      | Description    | Default value      |
| :---------------------- | :-------- | :-------------- | :------- |

> | `logLevel`     | `RSLogLevel`     | Controls how much of the log you want to see from the iOS SDK.                | `RSLogLevel.None`                                                           |

Reviewer: ```suggestion
| `logLevel`     | `RSLogLevel`     | Controls how much of the log you want to see from the SDK.                | `RSLogLevel.none`                                                           |
```

| `sleepTimeout`          | Integer     | Minimum waiting time to flush the events to the server.  | `10 seconds`    |
| `configRefreshInterval` | Integer     | The SDK fetches the config from dashboard after this specified time.    | `2 hours`  |
| `trackLifecycleEvents`  | Boolean | Determines if the SDK will capture application life cycle events automatically.    | `true`    |
| `recordScreenViews`     | Boolean | Determines if the SDK will capture will capture screen view events automatically.   | `false`   |
| `controlPlaneUrl`       | String  | This parameter should be changed **only if** you are self-hosting the control plane. Check the [Self-hosted control plane](https://rudderstack.com/docs/stream-sources/rudderstack-sdk-integration-guides/rudderstack-ios-sdk/#self-hosted-control-plane) section below for more information. The SDK will add `/sourceConfig` along with this URL to fetch the required configuration. | `https://api.rudderlabs.com`   |

### Self-hosted control plane

> If you are using a device mode destination like Adjust, Firebase, etc., the iOS SDK needs to fetch the required configuration from the control plane. If you are using the [Control Plane Lite](https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/) utility to host your own control plane, then follow [this section](https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/#control-plane-url) to specify `controlPlaneUrl` in `RudderConfig` that points to your hosted source configuration file.

Reviewer: ```suggestion
If you are using a device mode destination like Adjust, Firebase, etc., the SDK needs to fetch the required configuration from the control plane. If you are using the [Control Plane Lite](https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/) utility to host your own control plane, then follow [this section](https://www.rudderstack.com/docs/rudderstack-open-source/control-plane-lite/#control-plane-url) to specify `controlPlaneUrl` in `RSConfig` that points to your hosted source configuration file.
```

</div>

An sample `identify` call is shown below:

<Tabs>
  <TabList>
    <Tab>Objective-C</Tab>
    <Tab>Swift</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```objectivec
[[RSClient sharedInstance] identify:@"user_id" traits:@{@"email": @"alex@example.com"}];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
RSClient.sharedInstance().identify("user_id", traits: ["email": "alex@example.com"])
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

The `identify` method accepts the following parameters:

| Name      | Data type      | Presence | Description         |
| :-------- | :------------- | :------- | :--------------------------- |
| `userId`  | `NSString`     | Required      | Uniquely identifies the visiting user.   |
| `traits`  | `NSDictionary` | Optional       | Information on the user traits. Use the `dict` method of `RudderTraits` to convert to `NSDictionary` easily. |

> | `options` | `RudderOption` | Optional       | Extra options for the `identify` event.      |

Reviewer: ```suggestion
| `option` | `RSOption` | Optional       | Extra options for the `identify` event.      |
```

A sample `screen` event is as shown:

<Tabs>
  <TabList>
    <Tab>Objective-C</Tab>
    <Tab>Swift</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```objectivec
[client screen:@"ViewController" category:NULL properties:@{
  @"key_1": @"value_1",
  @"key_2": @"value_2"
} 
options:NULL];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
client.screen("ViewController", properties: [
    "key_1": "value_1",
    "key_2": "value_2"
])
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

The `screen` method accepts the following parameters:

| Name         | Data type      | Presence | Description                                                              |
| :----------- | :------------- | :------- | :--------------------------------------------- |
| `screenName` | `NSString`     | Required      | Name of the screen viewed by the user.     |
| `properties` | `NSDictionary` | Optional       | Extra property object to be passed along with the `screen` call. |

> | `options`    | `RudderOption` | Optional       | Extra options to be passed along with the `screen` call.     |

Author: should this be `option` or `options`? As per the spec shared by you [here](https://www.notion.so/rudderstacks/Swift-migration-changes-Core-SDK-5ab9a96f526344d4bde3a149723cd18b) (refer to the Screen section)

A sample `group` call is shown below:

<Tabs>
  <TabList>
    <Tab>Objective-C</Tab>
    <Tab>Swift</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```objectivec
[client group:@"sample_group_id" traits:@{
  @"key_1": @"value_1",
  @"key_2": @"value_2"
} 
options:NULL];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
client.group("sample_group_id", traits: [
  "key_1": "value_1",
  "key_2": "value_2"
])
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

The `group` method accepts the following parameters:

| Name      | Data type      | Presence | Description                |
| :-------- | :------------- | :------- | :------------------------------------ |
| `groupId` | `String`       | Required      | The unique identifier of the group with which you want to associate your user.        |
| `traits`  | `NSDictionary` | Optional       | Any other property of the organization you want to pass along with the call. |
| `options` | `RudderOption` | Optional       | Extra event-level options to be passed along with the `group` call.              |

<div class="warningBlock">

> The iOS SDK does not persist the group traits across the sessions.

Author: What is the reason for this? I think this information will be beneficial.

    </TabPanels>
</Tabs>

Alternatively, you can use the following method signature

| Name      | Data type      | Presence | Description                                     |
| :-------- | :------------- | :------- | :---------------------------------------------- |
| `newId`   | `String`       | Required      | The new `userId` you want to assign to the user. |
| `options` | `RudderOption` | Optional       | Event-level options.                              |

<div class="infoBlock">

RudderStack replaces the old <code class="inline-code">userId</code> with the <code class="inline-code">newUserId</code> and persists that identification across the sessions.
</div>

## Reset

You can use the `reset` method to clear the persisted user traits for the `identify` call. This is required for the user logout operation.

<Tabs>
  <TabList>
    <Tab>Objective-C</Tab>
    <Tab>Swift</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```objectivec
[client reset];
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```swift
client.reset()
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

## Enabling/disabling user tracking via the `setOptOutStatus` API (GDPR support)

> RudderStack gives the users (for example, an EU user) the ability to opt out of tracking any user activity until they give their consent. You can do this by leveraging RudderStack's `setOptOutStatus` API.

Author: Is the `setOptOutStatus` API available only from v2?

- Finally, add the `CustomDestination` with the RudderStack iOS SDK after its initialization, as shown:

<Tabs>
  <TabList>
    <Tab>Swift</Tab>
    <Tab>Objective-C</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```swift
client.addDestination(CustomDestination())
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```objectivec
#import "&lt;PROJECT_MODULE_NAME&gt;-Swift.h"

[client addDestination:[[CustomDestination alloc] init]];
```
</span>
      </TabPanel>
    </TabPanels>
</Tabs>

## FAQs

### I'm facing issues building with Carthage on XCode 12. What should I do?

If you're facing an issue with Carthage and XCode 12, you can follow [this workaround](https://github.com/Carthage/Carthage/blob/master/Documentation/Xcode12Workaround.md) suggested by the Carthage team.

### Does the iOS SDK support the tvOS and watchOS platforms?

Yes, the iOS SDK supports tvOS and watchOS platforms. Refer to the table below for the version details:

| Platform | Minimum SDK version supported |
| :--------| :----------------|
| [tvOS](https://developer.apple.com/tvos/) | 1.1.0 |
| [watchOS](https://developer.apple.com/watchos/) | 1.3.0 |

> ### How do I migrate from v1.0.2?

Author: Please review this FAQ and confirm if the **updated** parameter names remain the same

> ### Does the iOS SDK support the tvOS and watchOS platforms?

Reviewer: Add macOS here.

To migrate from an older version 1.0.2, update the usage of the following classes:

| Previous Name | Updated Name |
| :---------------| :---------------|
| `RudderClient` | `RSClient`       |
| `RudderConfig`| `RSConfig`     |
| `RudderConfigBuilder` | `RSConfigBuilder` |
| `RudderLogLevelDebug` | `RSLogLevelDebug` |

> ### How do I ensure the events tracked just before closing/backgrounding the app are sent immediately and not on the next app launch?

Author: If `enableBackgroundMode` is deprecated in this version, should we remove this FAQ as well?

> | `RudderConfigBuilder` | `RSConfigBuilder` |

Reviewer: `RSConfigBuilder` - this class is not available.
